File: src/main/java/io/boson/bson/Boson.java
package io.boson.bson;

import java.nio.ByteBuffer;
import java.util.concurrent.CompletableFuture;
import java.util.function.Consumer;
import java.util.function.Function;


public interface Boson {

    /**
     * Make an Extractor that will call the extract function (Consumer) according to
     * the given expression.
     * @param expression
     * @param extractFunction
     * @param <T>
     * @return a Boson that is a BosonExtractor
     */
    static <T> Boson extractor(String expression, Consumer<T> extractFunction) {
        // TODO construct an extractor
        return null;
    }

    /**
     * Make an Injector that will call the inject function (of T -> T) according to
     * the given expression.
     * @param expression
     * @param injectFunction
     * @param <T>
     * @return
     */
    static <T> Boson injector(String expression, Function<T,T> injectFunction) {
        // TODO construct an injector
        return null;
    }

    /**
     * Apply this Boson to the byte array that arrives and at some point in the future complete
     * the future with the resulting byte array. In the case of an Extractor this will result in
     * the immutable byte array being returned unmodified.
     * @param bsonByteEncoding
     * @return
     */
    CompletableFuture<byte []> go(final byte [] bsonByteEncoding);

    /**
     * Apply this Boson to the byte array that arrives and at some point in the future complete
     * the future with the resulting byte array. In the case of an Extractor tis will result in
     * the immutable byte array being returned unmodified.
     * @param bsonByteBufferEncoding
     * @param <T>
     * @return
     */
    CompletableFuture<ByteBuffer> go(final ByteBuffer bsonByteBufferEncoding);


    /**
     * Fuse one Boson to another. The boson that is this should be executed first before the
     * boson that is the parameter in teh case of update/read conflicts.
     * the immutable byte array being returned unmodified.
     * @param the Boson to fuse to.
     * @return the fused Boson
     */
    Boson fuse(final Boson boson);

}

File: src/main/scala/io/boson/bson/Boson.scala
package io.boson.bson

import java.nio.ByteBuffer
import java.util.function.Consumer
import java.util.concurrent.CompletableFuture

import io.boson.bson

trait Boson {

  class BosonEx(_expression: String, _extractFunction: CompletableFuture[_]) {
    val expression: String = this._expression
    val extractFunction: CompletableFuture[_] = this._extractFunction

  }

  object BosonEx {
    def extractor(expression: String, extractFunction: CompletableFuture[_]): io.boson.bson.bsonImpl.Boson = {  //  instantiates a BosonExtractor
      val bsonValidatedEncoded: Array[Byte] = Array(1,2,3)
      new io.boson.bson.bsonImpl.Boson(byteArray = Option(bsonValidatedEncoded))
    }
  }

  class BosonIj(_expression: String, _injectFunction: (Any) => Any) {
    val expression: String = this._expression
    val injectFunction: (Any) => Any = this._injectFunction
  }

  object BosonIj {
    def injector(_expression: String, _injectFunction: (Any) => Any): io.boson.bson.bsonImpl.Boson = {
      val bsonValidatedEncoded: Array[Byte] = Array(1,2,3)
      new io.boson.bson.bsonImpl.Boson(byteArray = Option(bsonValidatedEncoded))
    }
  }

  def go(bsonByteEncoding: Array[Byte]): CompletableFuture[Array[Byte]]

  def go(bsonByteBufferEncoding: ByteBuffer): CompletableFuture[ByteBuffer]

  def fuse(boson: io.boson.bson.bsonImpl.Boson): io.boson.bson.bsonImpl.Boson
}



object Driver extends Boson {

  import java.util.concurrent.CompletableFuture

  val result = new CompletableFuture[String]
  val bsonValidatedEncoded: Array[Byte] = Array(1,2,3)


  val bosonEx: BosonEx = new BosonEx("first", result)
  val boson: io.boson.bson.bsonImpl.Boson = BosonEx.extractor(bosonEx.expression, bosonEx.extractFunction)
  //  boson.go(bsonValidatedEncoded)

  val bosonIj: BosonIj = new BosonIj("first", _=> "Adeus")
  val boson2: io.boson.bson.bsonImpl.Boson = BosonIj.injector(bosonIj.expression, bosonIj.injectFunction)
  //val result: CompletableFuture[Array[Byte]] = boson.go(bsonValidatedEncoded)


  override def go(bsonByteEncoding: Array[Byte]) = ???

  override def go(bsonByteBufferEncoding: ByteBuffer) = ???

  override def fuse(boson: bsonImpl.Boson) = ???
}
